src/transformStateWithClones.js

```diff
@@ -7,7 +7,44 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultAllState = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        const { extraData } = action;
+
+        addProperties(newState, extraData);
+        break;
+
+      case 'removeProperties':
+        const { keysToRemove } = action;
```

Comment:
Consider adding validation to check if `keysToRemove` exists in the action object to prevent potential runtime errors.